python/lsst/obs/base/filters.py

```diff
@@ -0,0 +1,194 @@
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Classes to allow obs packages to define the filters used by an Instrument
+and for use by `lsst.afw.image.Filter`, gen2 dataIds, and gen3 Dimensions.
+"""
+
+__all__ = ("FilterDefinition", "FilterDefinitionCollection")
+
+import dataclasses
+import collections.abc
+
+import numpy as np
+
+import lsst.afw.image.utils
+
+
+class FilterDefinitionCollection(collections.abc.Sequence):
+    """An order-preserving collection of FilterDefinitions.
+
+    Parameters
+    ----------
+    filters : sequence
+        The filters in this collection.
+    """
+    class __Singleton:
+        """Store whether the `lsst.afw.image.Filter` singletons have been
+        defined, to prevent subtle errors if different filter collections
+        attempt to define themselves.
+        """
+        def __init__(self):
+            self.defined = None
+
+    _defined = None
+    """Whether these filters have been defined via
+    `~lsst.afw.image.utils.defineFilter`. If so, set to ``self`` to identify
+    the filter collection that defined them.
+    """
+
+    def __init__(self, *filters):
+        if FilterDefinitionCollection._defined is None:
+            FilterDefinitionCollection._defined = FilterDefinitionCollection.__Singleton()
+        self._filters = list(filters)
+
+    def __getitem__(self, key):
+        return self._filters[key]
+
+    def __len__(self):
+        return len(self._filters)
+
+    def __str__(self):
+        return "FilterDefinitions(" + ', '.join(str(f) for f in self._filters) + ')'
+
+    def defineFilters(self):
+        """Define all the filters to `lsst.afw.image.Filter`.
+
+        `~lsst.afw.image.Filter` objects are singletons, so we protect against
+        filters being defined multiple times.
+
+        Raises
+        ------
+        RuntimeError
+            Raised if any other `FilterDefinitionCollection` has already called
+            ``defineFilters``.
+        """
+        if self._defined.defined is None:
+            lsst.afw.image.utils.resetFilters()
+            for filter in sorted(self._filters):
+                filter.defineFilter()
+            self._defined.defined = self
+        elif self._defined.defined is self:
+            # noop: we've already defined these filters, so do nothing
+            pass
+        else:
+            msg = f"afw Filters were already defined on: {self._defined.defined}"
+            raise RuntimeError(msg)
+
+
+@dataclasses.dataclass(frozen=True, order=True)
```

Comment:
What's the use case for `order=True`?

Reply:
Hmmm... You'd mentioned that the collection had to be order preserving (for a reason I've now forgotten), so I think I thought that we would want to be able to have strict ordering of the filters. But that's not the same thing. I'll remove it.